pkg/tkctl/cmd/check/check.go

```diff
@@ -0,0 +1,479 @@
+			return fmt.Errorf("failed to decode tidb config: %v", err)
+		}
+		if tidbcfg.Log.Level != "info" {
+			warnPrintf("The recommended TiDB log-level is info, the current is %s\n", tidbcfg.Log.Level)
+		}
+		if tidbcfg.Binlog.IgnoreError {
+			warnPrintln("TiDB binlog.ignore-error is enabled")
+		}
+		if tidbcfg.PreparedPlanCache.Enabled {
+			warnPrintln("TiDB prepared-plan-cache is an experimental feature, not recommended for production use")
+		}
+		if tidbcfg.TxnLocalLatches.Enabled {
+			warnPrintln("TiDB txn-local-latches is for write conflict heavy business")
+		}
+	}
+	if cfg, exist := tikvConfigMap.Data["config-file"]; exist {
+		if o.Verbose {
+			infoPrintln("#### TiKV Config ####")
+			fmt.Println(cfg)
+			infoPrintln("#### TiKV Config ####\n")
+		}
+		_, err = toml.Decode(cfg, &tikvcfg)
+		if err != nil {
+			return fmt.Errorf("failed to decode tikv config: %v", err)
+		}
+		tikvResources := util.ResourceRequirement(tidbcluster.Spec.TiKV.ContainerSpec)
+		memoryLimit := tikvResources.Limits[corev1.ResourceMemory]
+		m, b1 := memoryLimit.AsInt64()
+		if b1 {
+			// rocksdb.writecf.block-cache-size 10%~30% of memory limit
+			// rocksdb.defaultcf.block-cache-size 30%~50% of memory limit
+			if int64(tikvcfg.Rocksdb.Writecf.BlockCacheSize) < m/10 || int64(tikvcfg.Rocksdb.Writecf.BlockCacheSize) > 3*m/10 {
+				errorPrintf("TiKV rocksdb.writecf.block-cache-size should be 10%% ~ 30%% of memory limit, but got %d/%d\n", tikvcfg.Rocksdb.Writecf.BlockCacheSize, m)
+			}
+			if int64(tikvcfg.Rocksdb.Defaultcf.BlockCacheSize) < 3*m/10 || int64(tikvcfg.Rocksdb.Defaultcf.BlockCacheSize) > m/2 {
+				errorPrintf("TiKV rocksdb.defaultcf.block-cache-size should be 30%% ~ 50%% of memory limit, but got %d/%d\n", tikvcfg.Rocksdb.Defaultcf.BlockCacheSize, m)
+			}
+			infoPrintf("TiKV rocksdb.lockcf.block-cache-size: %dMiB\n", tikvcfg.Rocksdb.Lockcf.BlockCacheSize/1024/1024)
+		}
+		if !tikvcfg.RaftStore.SyncLog {
+			errorPrintln("Raftstore sync-log is disabled")
+		}
+		if tikvcfg.LogLevel != "info" {
+			warnPrintf("The recommended TiKV log-level is info, the current is %s\n", tikvcfg.LogLevel)
+		}
+	}
+	if cfg, exist := pumpConfigMap.Data["pump-config"]; exist {
+		if o.Verbose {
+			infoPrintln("#### Pump Config ####")
+			fmt.Println(cfg)
+			infoPrintln("#### Pump Config ####\n")
+		}
+	}
+	if cfg, exist := pdConfigMap.Data["config-file"]; exist {
+		if o.Verbose {
+			infoPrintln("#### PD Config ####")
+			fmt.Println(cfg)
+			infoPrintln("#### PD Config ####\n")
+		}
+		_, err = toml.Decode(cfg, &pdcfg)
+		if err != nil {
+			return fmt.Errorf("failed to decode pd config: %v", err)
+		}
+		if pdcfg.Log.Level != "info" {
+			warnPrintf("The recommended PD log-level is info, the current is %s\n", pdcfg.Log.Level)
+		}
+		if pdcfg.Replication.MaxReplicas != 3 && pdcfg.Replication.MaxReplicas != 5 {
+			errorPrintf("Production cluster must set TiKV data replication factor to 3 or 5, the current is %d\n", pdcfg.Replication.MaxReplicas)
```

Comment:
should we need add the `max-replicas` to [pd config](https://github.com/pingcap/tidb-operator/blob/master/charts/tidb-cluster/values.yaml#L67-L68)? or the check will failed with:

```
Production cluster must set TiKV data replication factor to 3 or 5, the current is 0
```

Reply:
All the components have default configuration, I'll merge user-provided configuration with the default value and then do the check.